ansible_base/authentication/authenticator_plugins/azuread.py

```diff
@@ -74,6 +79,25 @@ def get_user_details(self, response):
         This method is an override from social-core/social_core/backends/azuread.py
         It allows us to control what the username is.
         """
+        # start with the assertion we got back
+        user_details = deepcopy(response)
+        # At this point the response has not been overlaid with the info in the id token
+        # so we need to get the info from the id token and overlay it
+        if 'access_token' in response and 'id_token' in response:
+            user_details.update(self.user_data(response['access_token'], response=response))
+        # Finally overlay the fields from what we want to return initially
         return_object = super().get_user_details(response)
-        return_object['username'] = response.get(self.setting("USERNAME_FIELD"), return_object['username'])
+        user_details.update(return_object)
+
+        if self.setting("USERNAME_FIELD"):
+            alternate_username = user_details.get(self.setting("USERNAME_FIELD"))
```

Comment:
Maybe trim this value?

Reply:
Like to the max length of username? (Which I think is 512)

Reply:
I don't know if we trim this anywhere else but it couldn't hurt.

Reply:
Oh I just meant to remove any stray whitespace before looking for the key.  But now that I think about it it's not a separate text box in the UI so probably not needed.